src/osp/tasks/top_execute.cpp

```diff
@@ -0,0 +1,169 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include "top_execute.h"
+#include "top_session.h"
+#include "top_worker.h"
+#include "execute_simple.h"
+
+#include "../logging.h"
+
+#include <Corrade/Containers/ArrayViewStl.h>
+
+#include <entt/core/any.hpp>
+
+#include <sstream>
+#include <vector>
+
+using Corrade::Containers::ArrayView;
+using Corrade::Containers::StridedArrayView2D;
+using Corrade::Containers::arrayView;
+
+namespace osp
+{
+
+void top_run_blocking(Tags const& tags, Tasks const& tasks, TopTaskDataVec_t& rTaskData, ArrayView<entt::any> topData, ExecutionContext& rExec)
+{
+    std::vector<entt::any> dataRefs;
+    WorkerContext context;
+
+    std::vector<uint64_t> tmpEnqueue(tags.m_tags.vec().size());
+
+    // Run until there's no tasks left to run
+    while (true)
+    {
+        std::vector<uint64_t> tasksToRun(tasks.m_tasks.vec().size());
+        task_list_available(tags, tasks, rExec, tasksToRun);
+        auto const tasksToRunBits = lgrn::bit_view(tasksToRun);
+        unsigned int const availableCount = tasksToRunBits.count();
+
+        if (availableCount == 0)
+        {
+            break;
+        }
+
+        // Choose first available task
+        auto const task = TaskId(*tasksToRunBits.ones().begin());
+
+        task_start(tags, tasks, rExec, task);
+
+        TopTask &rTopTask = rTaskData.m_taskData.at(std::size_t(task));
+        dataRefs.resize(rTopTask.m_dataUsed.size());
+        for (std::size_t i = 0; i < rTopTask.m_dataUsed.size(); ++i)
+        {
+            dataRefs.at(i) = topData[rTopTask.m_dataUsed[i]].as_ref();
+        }
+
+        TopTaskStatus const status = rTopTask.m_func(context, dataRefs);
+
+        BitSpan_t enqueue = (status == TopTaskStatus::Success) ? tmpEnqueue : BitSpan_t{};
+
+        task_finish(tags, tasks, rExec, task, enqueue);
+    }
+}
+
+void top_enqueue_quick(Tags const& tags, Tasks const& tasks, ExecutionContext& rExec, ArrayView<TagId const> tagsEnq)
+{
+    rExec.m_tagIncompleteCounts  .resize(tags.m_tags.capacity(), 0);
+    rExec.m_tagRunningCounts     .resize(tags.m_tags.capacity(), 0);
+    rExec.m_taskQueuedCounts     .resize(tasks.m_tasks.capacity(), 0);
+
+    std::vector<uint64_t> tagsToRun(tags.m_tags.vec().size());
+    to_bitspan(tagsEnq, tagsToRun);
+    task_enqueue(tags, tasks, rExec, tagsToRun);
+}
+
+static void check_task_dependencies(Tags const& tags, Tasks const& tasks, TopTaskDataVec_t const& taskData, std::vector<uint32_t> &rPath, uint32_t const task, bool& rGood)
```

Comment:
static functions aren't visible outside the cpp file they're defined in, but a lot of people get tripped up on that.

I recommend putting this into an anonymous namespace (as well as leaving it static) so that there's no ambiguity

Reply:
Is there really any risk here worth adding an extra pair of braces to keep track of? What exactly is this trip up?

Reply:
anonymous namespaces and static keywords are both signals to the reader about properties of the thing being annotated. Both is less easy to miss than only one.